### ISSUE TITLE
Add timeout handling to the WebSocket subscriber connections

### ISSUE THREAD

**Issue by DiegoDiverio** [channable/icepeak#104]
As discussed in https://github.com/channable/icepeak/pull/101, we don't actually handle timeouts for WebSocket connections on Icepeak's side at the moment. It could be remedied by checking if the subscriber actually responds with a pong when we send a ping.

**Comment by rkrzr** [comment]
> As discussed in https://github.com/channable/icepeak/pull/101, we don't actually handle timeouts for WebSocket connections on Icepeak's side at the moment.

Could you elaborate on this?

I think what we would want is for each message to be delivered within, say, 60 seconds max. If the client doesn't respond/confirm in that time, we should close the connection.

If there is no message to be delivered, we would want to periodically send a ping (and wait for a pong). But I though that the Haskell library already does this by default.

**Comment by DiegoDiverio** [comment]
At the moment, it works for *us*, because we have a reverse proxy mechanism that ensures this using `nginx`. It shall work for anyone who deploys using this type of architecture.

However, it is not done by default by the `websockets` library, even using `withPingThread`. The actual implementation has to be made manually.

**Comment by robbert-vdh** [comment]
This sounded like a low hanging fruit issue but it seems like the websockets library doesn't actually have a good API for handling pongs. The action passed to `withPingThread` gets called when the ping is sent, with no clean way to interact with the corresponding pong. The websocket connection does however have a `connectionOnPong` configuration option, so you can still hack this up by storing the timestamp when the last pong was received using an `MVar`, and then checking the interval between the last pong and the just sent ping in the ping thread's action. Could also be done the other way around by storing the ping and then checking if a pong occurred within a certain time frame using a timeout but that sounds a lot more complex. Any oppositions to implementing it this way, or suggestions for something better?

See https://github.com/jaspervdj/websockets/issues/159.

**Comment by DiegoDiverio** [comment]
If I remember correctly, @fatho proposed another way to do it than with the code mentioned in that link. Let me see if I can find it.

**Comment by fatho** [comment]
Sounds like a good approach. Just a small detail:

> `MVar POSIXTime`

Using the system clock for timeouts is not a good idea because DST changes, NTP syncs, and (though less frequent) leap seconds will all mess with your timeouts. DST shifts are probably the worst here because they could (depending on the actual implementation of the timeouts) make all timeouts expire immediately if the timeout is less than the DST shift amount.

It's better to use a monotonic clock, as e.g. provided by the [`clock`](https://hackage.haskell.org/package/clock-0.8.2) package (which we use in other projects too).

**Comment by robbert-vdh** [comment]
Yeah I assumed POSIXTime was monotonic, replaced it with the monotonic version of Clock when I realized it wasn't.
